### DISCUSSION TITLE
Better ways to marshal objects.

### DISCUSSION THREAD

**Discussion by dpeter99** [ClearFoundry/ClearScript#243] [Q&A]
Hi,
I'm trying to marshal a complex object to the JS side, which the user written scripts should be able to modify in-place in the most natural ways for the developer as possible.
An example I have is a List or Dictionary which seem to be mapped 1:1, and it doesn't seem to support the more standard JS ways of handling lists (for example, "listID" in myObject.List - for checking if an element is present in the Dictionary)
Is there any way to achieve 2-way marshalling of these data types to a JS native format?

**Comment by ClearScriptLib** [comment]
Hi @dpeter99,

Could you provide a code sample that doesn't work the way you expect? It's difficult to diagnose your issue without knowing what "listID" and "myObject" are.

ClearScript prefers not to convert between .NET and script objects because such conversion tends to be expensive and lossy. Instead it aims to provide full, easy access to both .NET and script objects from either side.

Thanks!

**Reply by dpeter99** [reply]
HI!
Sorry for the late response, some other projects came in.

Soo here is an example:
Let's say I have these classes
```cs
public class Project
{
    public string name { get; set; }
    
    public string author { get; set; }
    
    public Build build { get; set; }
}

public class Build
{
  public Dictionary<string,Step> steps = new();
}
```
And this JS function get's a project as the input
```js
function extensionC(project) {
    if(!("link" in project.build.steps)) {
        buildscript.Build.steps["link"] = {...};
    }
}
```

**Reply by ClearScriptLib** [reply]
HI @dpeter99,

The problem here is that C# and JavaScript indexing, despite their common syntax, are fundamentally different mechanisms.

In JavaScript, `steps["link"]` is 100% equivalent to `steps.link`, whereas in C# they're completely different. Furthermore, because they're equivalent in JavaScript, V8 doesn't let the host differentiate between them.

If you'd like to expose something that behaves like a plain JavaScript object, ClearScript gives you a number of options. First, you could use a ClearScript [property bag](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_PropertyBag.htm):

```C#
var bag = new PropertyBag();
bag["foo"] = 123;
bag["bar"] = 456.789;
```

Another possibility is to use an instance of [`ExpandoObject`](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject):

```C#
var expando = new ExpandoObject() as IDictionary<string, object>;
expando["foo"] = "baz";
expando["bar"] = "qux";
```

Finally, you could use an actual script object:

```C#
var scriptObject = engine.Evaluate("({})") as ScriptObject;
scriptObject["foo"] = true;
scriptObject["bar"] = DateTime.Now;
```

You could pass any of these to a script function that requires an object with `foo` and `bar` properties. For example:

```C#
engine.AddHostType(typeof(Console));
engine.Execute(@"
    function dump(arg) {
        if ('foo' in arg) {
            Console.WriteLine(arg['foo']);
        }
        if ('bar' in arg) {
            Console.WriteLine(arg.bar);
        }
    }
");
engine.Script.dump(bag);
engine.Script.dump(expando);
engine.Script.dump(scriptObject);
```

Cheers!
